File: projeto-pronto-para-git/services/javascript/src/application/catalogService.js
export class CatalogDomainError extends Error {
  constructor(status, code, message) {
    super(message);
    this.name = "CatalogDomainError";
    this.status = status;
    this.code = code;
  }
}

const INITIAL_USER_COUNT = 200;
const INITIAL_SONG_COUNT = 400;
const INITIAL_PLAYLIST_COUNT = 300;

const BASE_DATA = {
  users: [
    { id: "u1", name: "Maya Ribeiro", email: "maya.ribeiro@example.test" },
    { id: "u2", name: "Icaro Mendes", email: "icaro.mendes@example.test" },
    { id: "u3", name: "Luna Ferreira", email: "luna.ferreira@example.test" }
  ],
  songs: [
    {
      id: "s1",
      title: "Pulso da Serra",
      artist: "Coletivo Prisma",
      album: "Rotas Claras",
      durationSeconds: 214
    },
    {
      id: "s2",
      title: "Vento Circular",
      artist: "Duo Cobalto",
      album: "Mapa Sonoro",
      durationSeconds: 188
    },
    {
      id: "s3",
      title: "Rua das Nuvens",
      artist: "Nina Vale",
      album: "Entre Pontes",
      durationSeconds: 241
    },
    {
      id: "s4",
      title: "Tambor de Vidro",
      artist: "Banda Meridiano",
      album: "Plano B",
      durationSeconds: 203
    },
    {
      id: "s5",
      title: "Manha de Codigo",
      artist: "Quarteto Latitude",
      album: "Sprint Azul",
      durationSeconds: 267
    }
  ],
  playlists: [
    { id: "p1", userId: "u1", name: "Rotina Clara", songIds: ["s1", "s3", "s4"] },
    { id: "p2", userId: "u1", name: "Deploy", songIds: ["s2", "s5"] },
    { id: "p3", userId: "u2", name: "Estrada Nova", songIds: ["s1", "s2", "s4"] }
  ]
};

function generatedUsers() {
  const firstNames = [
    "Alice",
    "Bianca",
    "Caio",
    "Daniel",
    "Elisa",
    "Felipe",
    "Gabriela",
    "Heitor",
    "Isabela",
    "Jonas",
    "Laura",
    "Marcos",
    "Nina",
    "Otavio",
    "Paula",
    "Rafael",
    "Sofia",
    "Tiago",
    "Valeria",
    "Yuri"
  ];
  const surnames = [
    "Almeida",
    "Barbosa",
    "Campos",
    "Dias",
    "Freitas",
    "Gomes",
    "Lopes",
    "Martins",
    "Nogueira",
    "Pereira",
    "Rocha",
    "Santana"
  ];
  const users = [];
  for (let index = BASE_DATA.users.length + 1; index <= INITIAL_USER_COUNT; index += 1) {
    users.push({
      id: `u${index}`,
      name: `${firstNames[index % firstNames.length]} ${surnames[index % surnames.length]} ${String(index).padStart(3, "0")}`,
      email: `ouvinte${String(index).padStart(3, "0")}@example.com`
    });
  }
  return users;
}

function generatedSongs() {
  const titleWords = [
    "Aurora",
    "Cidade",
    "Circuito",
    "Delta",
    "Estrada",
    "Farol",
    "Jardim",
    "Litoral",
    "Neblina",
    "Pulso",
    "Ritmo",
    "Saturno",
    "Vento"
  ];
  const artists = [
    "Banda Central",
    "Duo Atlantico",
    "Grupo Prisma",
    "Lia e os Sinais",
    "Mar Aberto",
    "Norte Livre",
    "Projeto Beta",
    "Quarteto Solar",
    "Trio Estacao",
    "Vozes do Sul"
  ];
  const albums = ["Volume", "Sessao", "Arquivo", "Colecao", "Mapa", "Registro", "Caderno", "Ao Vivo"];
  const songs = [];
  for (let index = BASE_DATA.songs.length + 1; index <= INITIAL_SONG_COUNT; index += 1) {
    songs.push({
      id: `s${index}`,
      title: `${titleWords[index % titleWords.length]} ${String(index).padStart(3, "0")}`,
      artist: artists[index % artists.length],
      album: `${albums[index % albums.length]} ${String(1 + (index % 20)).padStart(2, "0")}`,
      durationSeconds: 120 + (index * 7) % 260
    });
  }
  return songs;
}

function generatedPlaylists() {
  const names = [
    "Trabalho",
    "Intervalo",
    "Treino",
    "Estrada",
    "Domingo",
    "Codigo",
    "Estudo",
    "Noite",
    "Cafe",
    "Viagem"
  ];
  const playlists = [];
  for (let index = BASE_DATA.playlists.length + 1; index <= INITIAL_PLAYLIST_COUNT; index += 1) {
    const userId = `u${1 + ((index * 5) % INITIAL_USER_COUNT)}`;
    const songIds = Array.from(
      { length: 5 },
      (_item, offset) => `s${1 + (((index * 7) + offset * 17) % INITIAL_SONG_COUNT)}`
    );
    playlists.push({
      id: `p${index}`,
      userId,
      name: `${names[index % names.length]} ${String(index).padStart(3, "0")}`,
      songIds
    });
  }
  return playlists;
}

function buildInitialData() {
  return {
    users: [...BASE_DATA.users.map((user) => ({ ...user })), ...generatedUsers()],
    songs: [...BASE_DATA.songs.map((song) => ({ ...song })), ...generatedSongs()],
    playlists: [
      ...BASE_DATA.playlists.map((playlist) => ({ ...playlist, songIds: [...playlist.songIds] })),
      ...generatedPlaylists()
    ]
  };
}

const INITIAL_DATA = buildInitialData();

function clone(value) {
  if (Array.isArray(value)) {
    return value.map((item) => clone(item));
  }
  if (value && typeof value === "object") {
    return Object.fromEntries(
      Object.entries(value).map(([key, item]) => [key, clone(item)])
    );
  }
  return value;
}

function requireString(value, fieldName) {
  if (typeof value !== "string" || value.trim() === "") {
    throw new CatalogDomainError(400, "INVALID_INPUT", `${fieldName} e obrigatorio`);
  }
  return value.trim();
}

function optionalString(value, fieldName) {
  if (value === undefined || value === null) {
    return null;
  }
  return requireString(value, fieldName);
}

function requirePositiveNumber(value, fieldName) {
  const number = Number.parseInt(value, 10);
  if (!Number.isFinite(number) || number <= 0) {
    throw new CatalogDomainError(400, "INVALID_INPUT", `${fieldName} deve ser positivo`);
  }
  return number;
}

function normalizeSongIds(songIds) {
  if (songIds === undefined || songIds === null) {
    return [];
  }
  if (typeof songIds === "string") {
    return songIds
      .split(",")
      .map((songId) => songId.trim())
      .filter(Boolean);
  }
  if (!Array.isArray(songIds)) {
    throw new CatalogDomainError(400, "INVALID_INPUT", "songIds deve ser lista ou texto separado por virgulas");
  }
  return songIds.map((songId) => requireString(songId, "songId"));
}

function toMap(items) {
  return new Map(items.map((item) => [item.id, clone(item)]));
}

function listValues(items) {
  return [...items.values()].map((item) => clone(item));
}

function listFilteredValues(items, predicate) {
  return [...items.values()].filter(predicate).map((item) => clone(item));
}

function plainTypeName(itemType) {
  return itemType.normalize("NFD").replace(/[\u0300-\u036f]/g, "");
}

export class CatalogService {
  constructor() {
    this.reset();
  }

  reset() {
    this.users = toMap(INITIAL_DATA.users);
    this.songs = toMap(INITIAL_DATA.songs);
    this.playlists = toMap(INITIAL_DATA.playlists);
    this.counters = {
      user: this.users.size + 1,
      song: this.songs.size + 1,
      playlist: this.playlists.size + 1
    };
    return { ok: true };
  }

  nextId(kind) {
    const prefixes = { user: "u", song: "s", playlist: "p" };
    const itemId = `${prefixes[kind]}${this.counters[kind]}`;
    this.counters[kind] += 1;
    return itemId;
  }

  assertExists(items, itemId, itemType) {
    if (!items.has(itemId)) {
      throw new CatalogDomainError(404, "NOT_FOUND", `${plainTypeName(itemType)} ${itemId} nao encontrado`);
    }
  }

  listUsers() {
    return listValues(this.users);
  }

  getUser(userId) {
    this.assertExists(this.users, userId, "Usuario");
    return clone(this.users.get(userId));
  }

  createUser(data = {}) {
    const user = {
      id: this.nextId("user"),
      name: requireString(data.name, "name"),
      email: requireString(data.email, "email")
    };
    this.users.set(user.id, user);
    return clone(user);
  }

  updateUser(userId, data = {}) {
    this.assertExists(this.users, userId, "Usuario");
    const current = this.users.get(userId);
    const updated = {
      ...current,
      name: optionalString(data.name, "name") || current.name,
      email: optionalString(data.email, "email") || current.email
    };
    this.users.set(userId, updated);
    return clone(updated);
  }

  deleteUser(userId) {
    this.assertExists(this.users, userId, "Usuario");
    this.users.delete(userId);
    this.playlists = new Map(
      [...this.playlists.entries()].filter(([, playlist]) => playlist.userId !== userId)
    );
    return { ok: true };
  }

  listSongs() {
    return listValues(this.songs);
  }

  getSong(songId) {
    this.assertExists(this.songs, songId, "Musica");
    return clone(this.songs.get(songId));
  }

  createSong(data = {}) {
    const song = {
      id: this.nextId("song"),
      title: requireString(data.title, "title"),
      artist: requireString(data.artist, "artist"),
      album: requireString(data.album, "album"),
      durationSeconds: requirePositiveNumber(data.durationSeconds, "durationSeconds")
    };
    this.songs.set(song.id, song);
    return clone(song);
  }

  updateSong(songId, data = {}) {
    this.assertExists(this.songs, songId, "Musica");
    const current = this.songs.get(songId);
    const updated = {
      ...current,
      title: optionalString(data.title, "title") || current.title,
      artist: optionalString(data.artist, "artist") || current.artist,
      album: optionalString(data.album, "album") || current.album,
      durationSeconds:
        data.durationSeconds === undefined || data.durationSeconds === null
          ? current.durationSeconds
          : requirePositiveNumber(data.durationSeconds, "durationSeconds")
    };
    this.songs.set(songId, updated);
    return clone(updated);
  }

  deleteSong(songId) {
    this.assertExists(this.songs, songId, "Musica");
    this.songs.delete(songId);
    for (const playlist of this.playlists.values()) {
      playlist.songIds = playlist.songIds.filter((currentId) => currentId !== songId);
    }
    return { ok: true };
  }

  listPlaylists(filters = {}) {
    const userId = filters.userId || null;
    const songId = filters.songId || null;

    if (userId) {
      this.assertExists(this.users, userId, "Usuario");
    }
    if (songId) {
      this.assertExists(this.songs, songId, "Musica");
    }

    return listFilteredValues(
      this.playlists,
      (playlist) =>
        (!userId || playlist.userId === userId) &&
        (!songId || playlist.songIds.includes(songId))
    );
  }

  getPlaylist(playlistId) {
    this.assertExists(this.playlists, playlistId, "Playlist");
    return clone(this.playlists.get(playlistId));
  }

  createPlaylist(data = {}) {
    const userId = requireString(data.userId, "userId");
    const songIds = normalizeSongIds(data.songIds);
    this.assertExists(this.users, userId, "Usuario");
    for (const songId of songIds) {
      this.assertExists(this.songs, songId, "Musica");
    }

    const playlist = {
      id: this.nextId("playlist"),
      userId,
      name: requireString(data.name, "name"),
      songIds
    };
    this.playlists.set(playlist.id, playlist);
    return clone(playlist);
  }

  updatePlaylist(playlistId, data = {}) {
    this.assertExists(this.playlists, playlistId, "Playlist");
    const current = this.playlists.get(playlistId);
    const userId = optionalString(data.userId, "userId") || current.userId;
    const songIds =
      data.songIds === undefined || data.songIds === null
        ? current.songIds
        : normalizeSongIds(data.songIds);

    this.assertExists(this.users, userId, "Usuario");
    for (const songId of songIds) {
      this.assertExists(this.songs, songId, "Musica");
    }

    const updated = {
      ...current,
      userId,
      name: optionalString(data.name, "name") || current.name,
      songIds
    };
    this.playlists.set(playlistId, updated);
    return clone(updated);
  }

  deletePlaylist(playlistId) {
    this.assertExists(this.playlists, playlistId, "Playlist");
    this.playlists.delete(playlistId);
    return { ok: true };
  }

  listUserPlaylists(userId) {
    this.assertExists(this.users, userId, "Usuario");
    return listFilteredValues(this.playlists, (playlist) => playlist.userId === userId);
  }

  listPlaylistSongs(playlistId) {
    this.assertExists(this.playlists, playlistId, "Playlist");
    return this.playlists
      .get(playlistId)
      .songIds.filter((songId) => this.songs.has(songId))
      .map((songId) => clone(this.songs.get(songId)));
  }

  listSongPlaylists(songId) {
    this.assertExists(this.songs, songId, "Musica");
    return listFilteredValues(this.playlists, (playlist) => playlist.songIds.includes(songId));
  }
}

export function plainError(error) {
  return {
    error: {
      code: error.code || "INTERNAL_ERROR",
      message: error.message || "Erro inesperado"
    }
  };
}


export const DomainError = CatalogDomainError;
export const MusicStore = CatalogService;

File: projeto-pronto-para-git/services/python/music_service/application/catalog_service.py
from __future__ import annotations

from copy import deepcopy
from threading import RLock


class CatalogDomainError(Exception):
    def __init__(self, status: int, code: str, message: str):
        super().__init__(message)
        self.status = status
        self.code = code


INITIAL_USER_COUNT = 200
INITIAL_SONG_COUNT = 400
INITIAL_PLAYLIST_COUNT = 300


BASE_DATA = {
    "users": [
        {"id": "u1", "name": "Maya Ribeiro", "email": "maya.ribeiro@example.test"},
        {"id": "u2", "name": "Icaro Mendes", "email": "icaro.mendes@example.test"},
        {"id": "u3", "name": "Luna Ferreira", "email": "luna.ferreira@example.test"},
    ],
    "songs": [
        {
            "id": "s1",
            "title": "Pulso da Serra",
            "artist": "Coletivo Prisma",
            "album": "Rotas Claras",
            "durationSeconds": 214,
        },
        {
            "id": "s2",
            "title": "Vento Circular",
            "artist": "Duo Cobalto",
            "album": "Mapa Sonoro",
            "durationSeconds": 188,
        },
        {
            "id": "s3",
            "title": "Rua das Nuvens",
            "artist": "Nina Vale",
            "album": "Entre Pontes",
            "durationSeconds": 241,
        },
        {
            "id": "s4",
            "title": "Tambor de Vidro",
            "artist": "Banda Meridiano",
            "album": "Plano B",
            "durationSeconds": 203,
        },
        {
            "id": "s5",
            "title": "Manha de Codigo",
            "artist": "Quarteto Latitude",
            "album": "Sprint Azul",
            "durationSeconds": 267,
        },
    ],
    "playlists": [
        {"id": "p1", "userId": "u1", "name": "Rotina Clara", "songIds": ["s1", "s3", "s4"]},
        {"id": "p2", "userId": "u1", "name": "Deploy", "songIds": ["s2", "s5"]},
        {"id": "p3", "userId": "u2", "name": "Estrada Nova", "songIds": ["s1", "s2", "s4"]},
    ],
}


def _generated_users():
    first_names = [
        "Alice",
        "Bianca",
        "Caio",
        "Daniel",
        "Elisa",
        "Felipe",
        "Gabriela",
        "Heitor",
        "Isabela",
        "Jonas",
        "Laura",
        "Marcos",
        "Nina",
        "Otavio",
        "Paula",
        "Rafael",
        "Sofia",
        "Tiago",
        "Valeria",
        "Yuri",
    ]
    surnames = [
        "Almeida",
        "Barbosa",
        "Campos",
        "Dias",
        "Freitas",
        "Gomes",
        "Lopes",
        "Martins",
        "Nogueira",
        "Pereira",
        "Rocha",
        "Santana",
    ]
    start = len(BASE_DATA["users"]) + 1
    return [
        {
            "id": f"u{index}",
            "name": f"{first_names[index % len(first_names)]} {surnames[index % len(surnames)]} {index:03d}",
            "email": f"ouvinte{index:03d}@example.com",
        }
        for index in range(start, INITIAL_USER_COUNT + 1)
    ]


def _generated_songs():
    title_words = [
        "Aurora",
        "Cidade",
        "Circuito",
        "Delta",
        "Estrada",
        "Farol",
        "Jardim",
        "Litoral",
        "Neblina",
        "Pulso",
        "Ritmo",
        "Saturno",
        "Vento",
    ]
    artists = [
        "Banda Central",
        "Duo Atlantico",
        "Grupo Prisma",
        "Lia e os Sinais",
        "Mar Aberto",
        "Norte Livre",
        "Projeto Beta",
        "Quarteto Solar",
        "Trio Estacao",
        "Vozes do Sul",
    ]
    albums = ["Volume", "Sessao", "Arquivo", "Colecao", "Mapa", "Registro", "Caderno", "Ao Vivo"]
    start = len(BASE_DATA["songs"]) + 1
    return [
        {
            "id": f"s{index}",
            "title": f"{title_words[index % len(title_words)]} {index:03d}",
            "artist": artists[index % len(artists)],
            "album": f"{albums[index % len(albums)]} {1 + (index % 20):02d}",
            "durationSeconds": 120 + (index * 7) % 260,
        }
        for index in range(start, INITIAL_SONG_COUNT + 1)
    ]


def _generated_playlists():
    names = [
        "Trabalho",
        "Intervalo",
        "Treino",
        "Estrada",
        "Domingo",
        "Codigo",
        "Estudo",
        "Noite",
        "Cafe",
        "Viagem",
    ]
    start = len(BASE_DATA["playlists"]) + 1
    playlists = []
    for index in range(start, INITIAL_PLAYLIST_COUNT + 1):
        user_id = f"u{1 + ((index * 5) % INITIAL_USER_COUNT)}"
        song_ids = [
            f"s{1 + (((index * 7) + offset * 17) % INITIAL_SONG_COUNT)}"
            for offset in range(5)
        ]
        playlists.append(
            {
                "id": f"p{index}",
                "userId": user_id,
                "name": f"{names[index % len(names)]} {index:03d}",
                "songIds": song_ids,
            }
        )
    return playlists


def _build_initial_data():
    return {
        "users": [*deepcopy(BASE_DATA["users"]), *_generated_users()],
        "songs": [*deepcopy(BASE_DATA["songs"]), *_generated_songs()],
        "playlists": [*deepcopy(BASE_DATA["playlists"]), *_generated_playlists()],
    }


INITIAL_DATA = _build_initial_data()


def _clone(value):
    if isinstance(value, dict):
        return {key: _clone(item) for key, item in value.items()}
    if isinstance(value, list):
        return [_clone(item) for item in value]
    return value


def _require_string(value, field_name: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise CatalogDomainError(400, "INVALID_INPUT", f"{field_name} e obrigatorio")
    return value.strip()


def _optional_string(value, field_name: str):
    if value is None:
        return None
    return _require_string(value, field_name)


def _require_positive_number(value, field_name: str) -> int:
    try:
        number = int(value)
    except (TypeError, ValueError):
        raise CatalogDomainError(400, "INVALID_INPUT", f"{field_name} deve ser positivo") from None

    if number <= 0:
        raise CatalogDomainError(400, "INVALID_INPUT", f"{field_name} deve ser positivo")
    return number


def _normalize_song_ids(song_ids) -> list[str]:
    if song_ids is None:
        return []
    if isinstance(song_ids, str):
        return [song_id.strip() for song_id in song_ids.split(",") if song_id.strip()]
    if not isinstance(song_ids, list):
        raise CatalogDomainError(400, "INVALID_INPUT", "songIds deve ser lista ou texto separado por virgulas")
    return [_require_string(song_id, "songId") for song_id in song_ids]


class CatalogService:
    def __init__(self):
        self._lock = RLock()
        self.reset()

    def reset(self):
        with self._lock:
            self.users = {item["id"]: _clone(item) for item in INITIAL_DATA["users"]}
            self.songs = {item["id"]: _clone(item) for item in INITIAL_DATA["songs"]}
            self.playlists = {item["id"]: _clone(item) for item in INITIAL_DATA["playlists"]}
            self.counters = {
                "user": len(self.users) + 1,
                "song": len(self.songs) + 1,
                "playlist": len(self.playlists) + 1,
            }
            return {"ok": True}

    def _next_id(self, kind: str) -> str:
        prefixes = {"user": "u", "song": "s", "playlist": "p"}
        item_id = f"{prefixes[kind]}{self.counters[kind]}"
        self.counters[kind] += 1
        return item_id

    @staticmethod
    def _list_values(items):
        return [_clone(item) for item in items.values()]

    @staticmethod
    def _assert_exists(items, item_id: str, item_type: str):
        if item_id not in items:
            raise CatalogDomainError(404, "NOT_FOUND", f"{item_type} {item_id} nao encontrado")

    def list_users(self):
        with self._lock:
            return self._list_values(self.users)

    def get_user(self, user_id: str):
        with self._lock:
            self._assert_exists(self.users, user_id, "Usuario")
            return _clone(self.users[user_id])

    def create_user(self, data):
        with self._lock:
            user = {
                "id": self._next_id("user"),
                "name": _require_string((data or {}).get("name"), "name"),
                "email": _require_string((data or {}).get("email"), "email"),
            }
            self.users[user["id"]] = user
            return _clone(user)

    def update_user(self, user_id: str, data):
        with self._lock:
            self._assert_exists(self.users, user_id, "Usuario")
            current = self.users[user_id]
            updated = {
                **current,
                "name": _optional_string((data or {}).get("name"), "name") or current["name"],
                "email": _optional_string((data or {}).get("email"), "email") or current["email"],
            }
            self.users[user_id] = updated
            return _clone(updated)

    def delete_user(self, user_id: str):
        with self._lock:
            self._assert_exists(self.users, user_id, "Usuario")
            del self.users[user_id]
            self.playlists = {
                playlist_id: playlist
                for playlist_id, playlist in self.playlists.items()
                if playlist["userId"] != user_id
            }
            return {"ok": True}

    def list_songs(self):
        with self._lock:
            return self._list_values(self.songs)

    def get_song(self, song_id: str):
        with self._lock:
            self._assert_exists(self.songs, song_id, "Musica")
            return _clone(self.songs[song_id])

    def create_song(self, data):
        with self._lock:
            song = {
                "id": self._next_id("song"),
                "title": _require_string((data or {}).get("title"), "title"),
                "artist": _require_string((data or {}).get("artist"), "artist"),
                "album": _require_string((data or {}).get("album"), "album"),
                "durationSeconds": _require_positive_number((data or {}).get("durationSeconds"), "durationSeconds"),
            }
            self.songs[song["id"]] = song
            return _clone(song)

    def update_song(self, song_id: str, data):
        with self._lock:
            self._assert_exists(self.songs, song_id, "Musica")
            current = self.songs[song_id]
            duration = (data or {}).get("durationSeconds")
            updated = {
                **current,
                "title": _optional_string((data or {}).get("title"), "title") or current["title"],
                "artist": _optional_string((data or {}).get("artist"), "artist") or current["artist"],
                "album": _optional_string((data or {}).get("album"), "album") or current["album"],
                "durationSeconds": current["durationSeconds"]
                if duration is None
                else _require_positive_number(duration, "durationSeconds"),
            }
            self.songs[song_id] = updated
            return _clone(updated)

    def delete_song(self, song_id: str):
        with self._lock:
            self._assert_exists(self.songs, song_id, "Musica")
            del self.songs[song_id]
            for playlist in self.playlists.values():
                playlist["songIds"] = [current_id for current_id in playlist["songIds"] if current_id != song_id]
            return {"ok": True}

    def list_playlists(self, filters=None):
        with self._lock:
            filters = filters or {}
            user_id = filters.get("userId") or None
            song_id = filters.get("songId") or None

            if user_id:
                self._assert_exists(self.users, user_id, "Usuario")
            if song_id:
                self._assert_exists(self.songs, song_id, "Musica")

            return [
                _clone(playlist)
                for playlist in self.playlists.values()
                if (not user_id or playlist["userId"] == user_id)
                and (not song_id or song_id in playlist["songIds"])
            ]

    def get_playlist(self, playlist_id: str):
        with self._lock:
            self._assert_exists(self.playlists, playlist_id, "Playlist")
            return _clone(self.playlists[playlist_id])

    def create_playlist(self, data):
        with self._lock:
            data = data or {}
            user_id = _require_string(data.get("userId"), "userId")
            song_ids = _normalize_song_ids(data.get("songIds"))
            self._assert_exists(self.users, user_id, "Usuario")
            for song_id in song_ids:
                self._assert_exists(self.songs, song_id, "Musica")

            playlist = {
                "id": self._next_id("playlist"),
                "userId": user_id,
                "name": _require_string(data.get("name"), "name"),
                "songIds": song_ids,
            }
            self.playlists[playlist["id"]] = playlist
            return _clone(playlist)

    def update_playlist(self, playlist_id: str, data):
        with self._lock:
            data = data or {}
            self._assert_exists(self.playlists, playlist_id, "Playlist")
            current = self.playlists[playlist_id]
            user_id = _optional_string(data.get("userId"), "userId") or current["userId"]
            song_ids = current["songIds"] if data.get("songIds") is None else _normalize_song_ids(data.get("songIds"))

            self._assert_exists(self.users, user_id, "Usuario")
            for song_id in song_ids:
                self._assert_exists(self.songs, song_id, "Musica")

            updated = {
                **current,
                "userId": user_id,
                "name": _optional_string(data.get("name"), "name") or current["name"],
                "songIds": song_ids,
            }
            self.playlists[playlist_id] = updated
            return _clone(updated)

    def delete_playlist(self, playlist_id: str):
        with self._lock:
            self._assert_exists(self.playlists, playlist_id, "Playlist")
            del self.playlists[playlist_id]
            return {"ok": True}

    def list_user_playlists(self, user_id: str):
        with self._lock:
            self._assert_exists(self.users, user_id, "Usuario")
            return [
                _clone(playlist)
                for playlist in self.playlists.values()
                if playlist["userId"] == user_id
            ]

    def list_playlist_songs(self, playlist_id: str):
        with self._lock:
            self._assert_exists(self.playlists, playlist_id, "Playlist")
            return [
                _clone(self.songs[song_id])
                for song_id in self.playlists[playlist_id]["songIds"]
                if song_id in self.songs
            ]

    def list_song_playlists(self, song_id: str):
        with self._lock:
            self._assert_exists(self.songs, song_id, "Musica")
            return [
                _clone(playlist)
                for playlist in self.playlists.values()
                if song_id in playlist["songIds"]
            ]


def plain_error(error: Exception):
    return {
        "error": {
            "code": getattr(error, "code", "INTERNAL_ERROR"),
            "message": str(error) or "Erro inesperado",
        }
    }


# Backward-compatible aliases used by legacy scripts and imports.
DomainError = CatalogDomainError
MusicStore = CatalogService

